zaino-state/src/chain_index/finalised_state.rs

```diff
@@ -1 +1,4383 @@
+    }
+
+    /// Clears stale reader slots by opening and closing a read transaction.
+    async fn clean_trailing(&self) -> Result<(), FinalisedStateError> {
+        let txn = self.env.begin_ro_txn()?;
+        drop(txn);
+        Ok(())
+    }
+
+    /// Opens an lmdb database if present else creates a new one.
+    async fn open_or_create_db(
+        env: &Environment,
+        name: &str,
+        flags: DatabaseFlags,
+    ) -> Result<Database, FinalisedStateError> {
+        match env.open_db(Some(name)) {
+            Ok(db) => Ok(db),
+            Err(lmdb::Error::NotFound) => env
+                .create_db(Some(name), flags)
+                .map_err(FinalisedStateError::LmdbError),
+            Err(e) => Err(FinalisedStateError::LmdbError(e)),
+        }
+    }
+
+    // *** DB write / delete methods ***
+    // These should only ever be used in a single DB control task.
+
+    /// Writes a given (finalised) [`ChainBlock`] to ZainoDB.
+    pub(crate) async fn write_block(&self, block: ChainBlock) -> Result<(), FinalisedStateError> {
+        self.status.store(StatusType::Syncing.into());
+        let block_hash = *block.index().hash();
+        let block_hash_bytes = block_hash.to_bytes()?;
+        let block_height = block.index().height().ok_or(FinalisedStateError::Custom(
+            "finalised state received non finalised block".to_string(),
+        ))?;
+        let block_height_bytes = block_height.to_bytes()?;
+
+        // check this is the *next* block in the chain.
+        tokio::task::block_in_place(|| {
+            let ro = self.env.begin_ro_txn()?;
+            let cur = ro.open_ro_cursor(self.headers)?;
+
+            // Position the cursor at the last header we currently have
+            match cur.get(None, None, lmdb_sys::MDB_LAST) {
+                // Database already has blocks
+                Ok((last_height_bytes, _last_header_bytes)) => {
+                    let last_height = Height::from_bytes(
+                        last_height_bytes.expect("Height is always some in the finalised state"),
+                    )?;
+
+                    // Height must be exactly +1 over the current tip
+                    if block_height.0 != last_height.0 + 1 {
+                        return Err(FinalisedStateError::Custom(format!(
+                            "cannot write block at height {:?}; \
+                     current tip is {:?}",
+                            block_height, last_height
+                        )));
+                    }
+                }
+                // no block in db, this must be genesis block.
+                Err(lmdb::Error::NotFound) => {
+                    if block_height.0 != 1 {
+                        return Err(FinalisedStateError::Custom(format!(
+                            "first block must be height 1, got {:?}",
+                            block_height
+                        )));
+                    }
```

Comment:
```suggestion
                    // Height must be exactly +1 over the current tip
                    if block_height.0 != last_height.0 + 1 {
                        return Err(FinalisedStateError::Custom(format!(
                            "cannot write block at height {block_height:?}; \
                     current tip is {last_height:?}"
                        )));
                    }
                }
                // no block in db, this must be genesis block.
                Err(lmdb::Error::NotFound) => {
                    if block_height.0 != 1 {
                        return Err(FinalisedStateError::Custom(format!(
                            "first block must be height 1, got {block_height:?}"
                        )));
                    }
```

clippy warning suggests -^

Reply:
this is weird I fixed all the warnings I get and i cant see any actual changes in this code..